File: Green-Ampt Infiltration/GreenAmpt.m
function [ t ] = GreenAmpt( Ks,sf,H,z,thetas,theta0 )
%GREENAMPT 此处显示有关此函数的摘要
%   此处显示详细说明
L=length(z);
t=zeros(L,1);
for i=1:L
    t(i)=(thetas-theta0)/Ks*(z(i)-(sf+H)*log((z(i)+sf+H)/(sf+H)));
end
end

File: Infiltration/Green-Ampt Infiltration/Conventional/Green_Ampt.cpp
#include <math>

/*
* Parameters:
* layer     : the water layer on the surface of the soil
* thetas    : the saturation of the soil
* theta0    : the initial water content of the soil
* Ksat      : the hydraulic conductivity of the soil
* sf        : shui xi li
* z         : the depth of the wet front 
*
* Equation:
* t = (thetas - theta0) / K * 
* (z - (sf + H） * ln((z + sf + H) / (sf + H ))
*
*/

inline double F(const double& z, const double& A, const double& B)
{
    return z - A * log(z / A + 1) - B;
}

inline double dF(const double&z, const double& A)
{
    return A / (z + A);
}

//Newton method is used to solve the Green-Ampt equation
double Green_Ampt(const double& Ks, const double& sf,
const double& H, const double& thetas, const double& theta0,
const double& t)
{
    double z = 0.0;

    double A = (sf + H);
    double B = (thetas - theta0) / Ks;

    while(true)
    {
        double norm = F(z, A, B) / dF(z, A);

        if(abs(norm) < 1e-6)
        {
            return z;
        }

        z -= norm;
    }
}